diff --git a/pcem-master/src/wx-sdl2-video-renderer.c b/pcem/src/wx-sdl2-video-renderer.c
index c06d64c..e00d7ce 100644
--- a/pcem-master/src/wx-sdl2-video-renderer.c
+++ b/pcem/src/wx-sdl2-video-renderer.c
@@ -69,7 +69,7 @@ void sdl_video_renderer_close()
         }
 }
 
-void sdl_video_renderer_update(SDL_Window* window, SDL_Rect updated_rect, BITMAP* screen)
+void sdl_video_renderer_update(SDL_Window* window, SDL_Rect updated_rect, PCEMBITMAP* screen)
 {
         SDL_UpdateTexture(texture, &updated_rect, &((uint32_t*) screen->dat)[updated_rect.y * screen->w + updated_rect.x], screen->w * 4);
 }
