diff --git a/pcem-master/src/wx-sdl2-video.h b/pcem/src/wx-sdl2-video.h
index a32d97d..ae7d448 100644
--- a/pcem-master/src/wx-sdl2-video.h
+++ b/pcem/src/wx-sdl2-video.h
@@ -11,7 +11,7 @@ struct sdl_render_driver;
 typedef struct sdl_renderer_t {
         int (*init)(SDL_Window* window, struct sdl_render_driver driver, SDL_Rect screen);
         void (*close)();
-        void (*update)(SDL_Window* window, SDL_Rect updated_rect, BITMAP* screen);
+        void (*update)(SDL_Window* window, SDL_Rect updated_rect, PCEMBITMAP* screen);
         void (*present)(SDL_Window* window, SDL_Rect texture_rect, SDL_Rect window_rect, SDL_Rect screen);
         int always_update;
 } sdl_renderer_t;
